openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/startup/startup-controller.service.ts

```diff
@@ -0,0 +1,43 @@
+import { Injectable, Injector } from '@angular/core';
+import { startupSequence } from './startup-sequence';
+
+export type StartupControllerState = true | 'in-progress' | false | 'error';
+
+export interface StartupControllerStatus {
+    completed: StartupControllerState;
+    error?: any;
+}
+
+@Injectable({
+    providedIn: 'root'
+})
+export class StartupController {
+    get status(): StartupControllerStatus {
+        return this._status;
+    }
+
+    private _status: StartupControllerStatus = {
+        completed: false
+    };
+
+    constructor(private _injector: Injector) { }
+
+    async beginStartupSequence(): Promise<StartupControllerStatus> {
+        if (this._status.completed !== false) {
+            throw new Error('startup sequence already initiated');
+        }
+
+        this._status = { completed: 'in-progress' };
```

Comment:
This might just be preference but I'm thinking why initialize the property above if you'll overwrite it with a new object instead of just changing the status
```suggestion
        this._status.completed = 'in-progress';
```

Reply:
I actually prefer the state as one whole immutable object (which I failed to actually do). Changes to to the object should be detected as a whole not as two seperate changes. It's also a little bit for safety. Changing the state should force you to intentionally attach an error or not.